type/__clean_path/gencode-remote

```diff
@@ -16,36 +16,18 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with cdist. If not, see <http://www.gnu.org/licenses/>.
-#
-
-[ ! -s "$__object/explorer/list" ] && exit 0
 
-if [ -f "$__object/parameter/path" ]
+if [ ! -s "$__object/explorer/list" ]
 then
-    path="$( cat "$__object/parameter/path" )"
-else
-    path="/$__object_id"
-fi
-
-pattern="$( cat "$__object/parameter/pattern" )"
-
-if [ -f "$__object/parameter/exclude" ]
-then
-    exclude="$( cat "$__object/parameter/exclude" )"
-
-    echo "find '$path' -mindepth 1 -maxdepth 1 -regex '$pattern'" \
-        "-and -not -regex '$exclude'" \
-        '-exec rm -rf {} \;'
-else
-    echo "find '$path' -mindepth 1 -maxdepth 1 -regex '$pattern'" \
-        '-exec rm -rf {} \;'
+    exit 0
 fi
 
-while read -r f
+while read -r remove_file
 do
-    echo "removed '$f'" >> "$__messages_out"
+    echo "rm -rf '$remove_file'"
```

Comment:
If a lot of files need to be deleted this can generate a lot of output.

So my thought was:
```sh
echo 'xargs rm -rf <"${__object:?}/explorer/list"'
```

This would generate a lot less `code-remote`, albeit at the cost of the expressiveness of `skonfig -d`.

What do you think?

It does not get rid of the timing issue, though.
So, maybe instead:
```sh
echo '"${__object:?}/explorer/list" | xargs rm -rf'
```

Reply:
This doesn't work because, for example:
`__object=/var/lib/skonfig/object/__clean_path/tmp/__clean_path_test/.cdist-kzy9wk3w`

But `explorer/list` is in `/var/lib/skonfig/conf/type/__clean_path/explorer/list`.

Reply:
Ah, yes, of course :facepalm: 

It should be:
```sh
echo '"${__type_explorer:?}/list" | xargs rm -f'
```

Reply:
But then we run `find` twice.

Which is more expensive - huge `gencode-remote` output or running `find` twice in target?

Also, another interesting idea: let's add arguments to type explorer...

In the beginning of explorer:
```
set -- '!' '-path' "$path" "$@"
```

Then we build find arguments based on parameters by prepending, for example:
```
set -- '-name' "$rm_name" "$@"
```

And last line is:
```
find "$path" "$@"
```

:thinking:

Reply:
Running `find` twice is cheaper because of no network overhead and VFS cache?

Reply:
Interesting thoughts.

I had the idea of manipulating the explorer's behaviour based on environment variables yesterday, but your approach of simply using `"$@"` is much better. I really like it!


> Running `find` twice is cheaper because of no network overhead and VFS cache?

Now that I'm thinking about it it makes sense. `find` is written in C and runs fast by directly iterating over the file system metadata.
If you then combine it with `-exec rm -rf {} +` you can also delete many files with only starting a few `rm(1)` processes.

Trying to pre-generate the list in the explorer runs `find`, copies the output across the network, stores it in a file on the config host, then reads it again, prepends `rm -rf` and writes it to the `code-remote` file again, then copies the file to the target and reads it one more time, to then execute one `rm(1)` process for each and every file.


My suggestion would be to only check if there are any files to be deleted in the explorer.
I think `find "$@" | read _` should work for to abort after the first result.

If we can somehow generate messages from one of the code scripts, we can re-run the full find with
`find ... -print -exec rm ... | sed 's/^/removed /' >$__messages_out`.

What do you think?

Reply:
Damn it, `$__type_explorer` is not available in `gencode-remote`.

Reply:
We should rethink the availability of the environment variables in the different scripts.
It confuses me all the time.

I get it, in an ideal world the `code-*` should know everything it needs to and not rely on explorers etc, but in practice this is just annoying.